mne/beamformer/_dics.py

```diff
@@ -1,32 +1,482 @@
+    .. [2] Sekihara & Nagarajan. Adaptive spatial filters for electromagnetic
+           brain imaging (2008) Springer Science & Business Media
+    .. [3] Hipp JF, Engel AK, Siegel M (2011) Oscillatory Synchronization
+           in Large-Scale Cortical Networks Predicts Perception.
+           Neuron 69:387-396.
+    """  # noqa: E501
+    allowed_ori = [None, 'normal', 'max-power']
+    if pick_ori not in allowed_ori:
+        raise ValueError('"pick_ori" should be one of %s.' % allowed_ori)
+
+    frequencies = [np.mean(freq_bin) for freq_bin in csd.frequencies]
+    n_freqs = len(frequencies)
+    n_orient = forward['sol']['ncol'] // forward['nsource']
+
+    picks = _setup_picks(info=info, forward=forward)
+    _, ch_names, proj, vertices, G = _prepare_beamformer_input(
+        info, forward, label, picks=picks, pick_ori=pick_ori,
+        leadfield_norm=leadfield_norm,
+    )
+    csd_picks = [csd.names.index(ch) for ch in ch_names]
+
+    n_sources = G.shape[1] // n_orient
+
+    logger.info('Computing DICS spatial filters...')
+    Ws = []
+    for i, freq in enumerate(frequencies):
+        if n_freqs > 1:
+            logger.info('    computing DICS spatial filter at %sHz (%d/%d)' %
+                        (freq, i + 1, n_freqs))
+
+        Cm = csd.get_matrix(index=i)
+
+        if real_filter:
+            Cm = Cm.real
+
+        # Ensure the CSD is in the same order as the leadfield
+        Cm = Cm[csd_picks, :][:, csd_picks]
+
+        # Tikhonov regularization using reg parameter to control for
+        # trade-off between spatial resolution and noise sensitivity
+        # eq. 25 in Gross and Ioannides, 1999 Phys. Med. Biol. 44 2081
+        Cm_inv, _ = _reg_pinv(Cm, reg, rcond='auto')
+
+        # Compute spatial filters
+        W = np.dot(G.T, Cm_inv)
+
+        for k in range(n_sources):
+            Wk = W[n_orient * k: n_orient * k + n_orient]
+            Gk = G[:, n_orient * k: n_orient * k + n_orient]
+
+            # Compute power at the source
+            Ck = np.dot(Wk, Gk)
+
+            # Normalize the spatial filters
+            if Wk.ndim == 2 and len(Wk) > 1:
+                # Free source orientation
+                # Officially, the formula calls for inv(Ck) @ Wk.
+                # However, since Ck is most likely rank deficient, the
+                # following is more stable in practice.
+                Wk /= np.diag(Ck)[:, np.newaxis]
+            else:
+                # Fixed source orientation
+                Wk /= Ck
+
+            if pick_ori == 'max-power':
+                # Compute spectral power by applying the spatial filters to the
+                # CSD matrix.
+                power = Wk.dot(Cm).dot(Wk.T)
+
+                # Compute the direction of max power
+                u, s, _ = np.linalg.svd(power)
+                max_power_ori = u[:, 0].real
+                max_power_ori /= np.linalg.norm(max_power_ori)
```

Comment:
I don't understand this bit. Aren't the rows/columns of U per definition orthonormal, so you are dividing by 1.0?

Reply:
When using eigenvectors, yes, but not when using SVD.
The reason we use SVD here is that `power` is often only approximately symmetric (or in our case Hermitian, since complex numbers are involved), due to the rank deduced pseudo-inverse tricks.

Reply:
When using eigenvectors, yes, but not when using SVD.

In a perfect world, `power` is a symmetric (or in this case Hermitian, since complex numbers are involved) matrix, in which case the eigenvectors and SVD components are the same and orthonormal.

However, we find that `power` is often only approximately symmetric, due to the rank deduced pseudo-inverse tricks we have to do. Now there is a choice between using eigenvectors and SVD components. The original Gross paper used SVD and this part of his original code survives to this day in Jan Kujala's pipeline. I've played around with using eigenvectors instead, got a very similar but slightly different result, couldn't decide which was better, so in the end chose to keep using SVD to reproduce the existing pipeline.

SVD will produce two sets of components which are orthogonal, but not normal as a rule. Hence the normalization step.

Reply:
I'm not sure there's any matrix property that would result in non-normal singular vectors, unless there is something spooky with the Python implementation of SVD. Can you provide an example matrix that would result in something like that?

Yes, I've also seen that eigenvalue decomposition and SVD can produce different results when calculating orientations under certain circumstances. No worries there, though we should (eventually) empirically test which implementation produces better results against brute-force selection the best orientation.

Reply:
I was confused by this statement at first, too -- but looking closer, the `.real` step (discarding the imaginary part) above, assuming that the singular vectors are actually meaningfully complex, would make them no longer normalized.

Reply:
@larsoner Aha, that makes sense. Though I wonder if it might make more sense (and yield better results) if the real part of `power` was taken first. (FWIW, that's the FieldTrip approach.)